cpp/src/strings/convert/convert_urls.cu

```diff
@@ -185,137 +192,184 @@ constexpr bool is_hex_digit(char ch)
+{
+  return (ptr[0] == '%' && is_hex_digit(ptr[1]) && is_hex_digit(ptr[2]));
+}
 
 /**
- * @brief Functor for detecting escape sequence positions that cross a string boundary.
+ * @brief Count the number of characters of each string after URL decoding.
+ *
+ * @tparam threadblock_size Number of threads in a threadblock. This template argument must be a
+ * multiple of `cudf::detail::warp_size`.
+ * @tparam block_size Number of characters which will be loaded into the shared memory at a time.
+ *
+ * @param[in] in_chars Character buffer for the input string column.
+ * @param[in] in_offsets Offset value of each string associated with `in_chars`.
+ * @param[out] out_counts Number of characters in each decode URL.
+ * @param[in] num_strings Number of strings to count the number of decoded characters.
  */
-struct url_decode_esc_position_filter {
-  device_span<int32_t const> const d_offsets{};
-
-  /**
-   * @brief Detects if an escape sequence crosses a string boundary
-   *
-   * Returns true for an escape sequence that straddles a boundary between two strings.
-   *
-   * @param esc_pos_idx Character position corresponding to the start of an escape sequence
-   * @return true The escape sequence crosses a string boundary
-   * @return false The escape sequence does not cross a string boundary
-   */
-  __device__ bool operator()(size_type esc_pos_idx) const
-  {
-    // find the end offset of the current string
-    size_type const* offset_ptr =
-      thrust::upper_bound(thrust::seq, d_offsets.begin(), d_offsets.end(), esc_pos_idx);
-    return esc_pos_idx + 2 >= *offset_ptr;
+template <int threadblock_size, int block_size>
+__global__ void url_decode_char_counter(char const* const in_chars,
+                                        offset_type const* const in_offsets,
+                                        offset_type* const out_counts,
+                                        size_type const num_strings)
+{
+  constexpr int warps_per_threadblock = threadblock_size / cudf::detail::warp_size;
+  __shared__ char temporary_buffer[warps_per_threadblock][block_size + 2];
+  __shared__ typename cub::WarpReduce<int8_t>::TempStorage cub_storage[warps_per_threadblock];
+
+  int global_thread_id  = blockIdx.x * blockDim.x + threadIdx.x;
+  int global_warp_id    = global_thread_id / cudf::detail::warp_size;
+  int local_warp_id     = threadIdx.x / cudf::detail::warp_size;
+  int warp_lane         = threadIdx.x % cudf::detail::warp_size;
+  int nwarps            = gridDim.x * blockDim.x / cudf::detail::warp_size;
+  char* in_chars_shared = temporary_buffer[local_warp_id];
+
+  // Loop through strings, and assign each string to a warp.
+  for (size_type row_idx = global_warp_id; row_idx < num_strings; row_idx += nwarps) {
+    auto in_chars_string          = in_chars + in_offsets[row_idx];
+    auto string_length            = in_offsets[row_idx + 1] - in_offsets[row_idx];
+    int nblocks                   = (string_length + block_size - 1) / block_size;
```

Comment:
Why do you need to add `block_size - 1` in this calculation? Is it because of the halo 0s? 

Either way, maybe better to define a `constexpr size_t HALO_SIZE = 2` at the top and use it everywhere that you have a raw constant `2` in this code.

Reply:
No, I am trying to calculate the ceiling of `string_length / block_size`. I figure this is a common operation. Do we have a helper function like that in cuDF? Maybe worth refactor this to an inline function as well.

Assign `halo_size` to a variable in https://github.com/rapidsai/cudf/pull/8622/commits/161468fdf6c4415d7990f887e8429c9e34800ec0.

Reply:
https://github.com/rapidsai/cudf/blob/7032f47b7456ac65077cda9ae614de0b58b26781/cpp/include/cudf/detail/utilities/integer_utils.hpp#L76-L79

Reply:
Ah of course you need the ceiling, that should have been obvious. We have a function like that [in cuIO](https://github.com/rapidsai/cudf/blob/be25a30ca20f3135f341c51b36cb075b376d5def/cpp/src/io/utilities/parsing_utils.cu#L23), @vuule would it make sense to move `divCeil` into a common header somewhere? It's such a simple operation that I could see an argument for just redefining again locally here rather than bringing in another header and increasing compile-time, and realistically most devs will probably not bother looking for such a function before rolling their own one-liner anyway.

+1 for `halo_size`.

Reply:
Hmm well that seems not great if we already have two essentially nearly identical implementations, the one in `integer_utils.hpp` seems slightly preferable though since it accepts const refs and heterogenous types. Maybe something that cuIO should be using instead of `divCeil`?

Reply:
Thanks David and Vyas for the suggestions! I switched to use `div_rounding_up_unsafe` instead of rolling my own in https://github.com/rapidsai/cudf/pull/8622/commits/afcb71221f647940ea41a19eed06f2497326c2ab.